src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -106,13 +105,12 @@ protected override void OnInit()
         // called by BufferedFileStreamStrategy
         internal override void OnBufferAllocated(byte[] buffer)
         {
-            Debug.Assert(_preallocatedOverlapped == null);
+            Debug.Assert(_reusableValueTaskSource == null);
 
-            _preallocatedOverlapped = new PreAllocatedOverlapped(ValueTaskSource.s_ioCallback, this, buffer);
+            _reusableValueTaskSource = new ValueTaskSource(this, buffer);
         }
 
-        private ValueTaskSource? CompareExchangeCurrentOverlappedOwner(ValueTaskSource? newSource, ValueTaskSource? existingSource)
-            => Interlocked.CompareExchange(ref _currentOverlappedOwner, newSource, existingSource);
+        private void TryToReuse(ValueTaskSource source) => Interlocked.CompareExchange(ref _reusableValueTaskSource, source, null);
```

Comment:
Doesn't this need to dispose the instance if it fails?
```suggestion
        private void TryToReuse(ValueTaskSource source)
        {
            source.Reset(); // will need to call ManualResetValueTaskSourceCore.Reset()
            if (Interlocked.CompareExchange(ref _reusableValueTaskSource, source, null) is not null)
            {
                source.Dispose();
            }
        }
```
?